### PR TITLE
Remove find_reasonable_step_size

### PR DESCRIPTION
**Pull request by fehiepsi**
Addresses #543. This PR drops the usage of `find_reasonable_step_size` in `hmc`.

### Reason

In HMC/NUTS, we only compile potential_fn and its grad 1 time: at leap_frog step. However, in find_reasonable_step_size, we compile potential_fn 2 times (to get initial potential energy and in the while-loop to find reasonable step_size) and its grad 1 time (in the while-loop to find reasonable step_size). Those together require JAX to compile a computational graph which is 2.5x larger than if we drop `find_reasonable_step_size`. It is known that JAX compiling time is superlinear w.r.t. its computational graph, so if the model is slow to compile, our MCMC will take a very long time for its initial step. Some evidence:
+ As @jhmenke observed in #543, removing this functionality reduces compiling time from 3m to 11s.
+ In our early development, we observed the slowness and introduced `heuristic_step_size` flag at hmc_init (see https://github.com/pyro-ppl/numpyro/pull/127). But we removed it at 0.2 release to clean up the interface, and the compiling time is acceptable for models we tested.
+ `hmm` example: 39s -> 28s elapsed time
+ `ode` example: 20s compiling -> 10s compiling

With this change, the compiling time of `sample_kernel` will be aligned with the compiling time of `value_and_grad(potential_fn)`.

### Effect

I used the following code to test the effect of removing this feature. Surprisingly, it seems that dual_averaging can do a pretty good job to remedy it
```python
from numpyro.infer.hmc_util import dual_averaging
import jax.numpy as np

ss_init, ss_update = dual_averaging()

ss_state = ss_init()
for i in range(10):
    accept_prob = 0.4  # 0.4, 0.6, 0.9
    ss_state = ss_update(0.8 - accept_prob, ss_state)
    print("current ss", np.exp(ss_state[0]), "    \taverage ss", np.exp(ss_state[1]))
```
With `accept_prob = 0.4`, we get
```
current ss 0.48322508     	average ss 0.48322508
current ss 0.15173525     	average ss 0.24267578
current ss 0.04085883     	average ss 0.111069076
current ss 0.010343172     	average ss 0.047984514
current ss 0.0025725756     	average ss 0.020001529
current ss 0.0006435767     	average ss 0.008161279
current ss 0.00016403742     	average ss 0.0032920958
current ss 4.290037e-05     	average ss 0.0013218971
current ss 1.1554674e-05     	average ss 0.0005309463
current ss 3.2104135e-06     	average ss 0.0002140629
```
we can see that if things are bad, step_size will drop pretty fast. On the other hand `accept_prob = 1`, we get
```
current ss 1.438551     	average ss 1.438551
current ss 2.5671825     	average ss 2.0299568
current ss 4.947172     	average ss 3.0005677
current ss 9.832708     	average ss 4.565091
current ss 19.71587     	average ss 7.0707974
current ss 39.418476     	average ss 11.069319
current ss 78.07797     	average ss 17.42865
current ss 152.67555     	average ss 27.504337
current ss 294.18542     	average ss 43.3985
current ss 558.1096     	average ss 68.34855
```
which shows that step_size (and its average) will be increased quickly if things look good. The length of the last window in warmup_adapter is 50, so `step_size` has the whole room to adjust in this interval. So I believe that inference results won't change if we remove this functionality.

As an example, running `hmm`, I don't observe differences in the inference result.

### CODE REVIEW

**Comment by neerajprad** [comment]
Always good to see code being simplified! My concern is that there might be cases where we are wasting adaptation steps in getting to a reasonable value of step size, and we may now start giving bad results on models that were previously working (though given your observations, that seems unlikely). Longer term, I think this is probably the right solution anyways since running longer adaptation will be faster than paying for repeated compilation time. In the meantime, a safer alternative would be to gate this behind a `find_heuristic_step_size=False` argument, which by default will not use the `find_reasonable_step_size` heuristic. If we don't notice any issues with the new default, we can later remove it completely  and simplify the interface. WDYT?

**Comment by fehiepsi** [comment]
> n the meantime, a safer alternative would be to gate 

yup, this is safer. Your concern is quite valid! I hope that in that case, we can replace it by other heuristic functions, which do not depend on potential_fn (maybe just using the norm of the last inverse_mm and the current inverse_mm) so that we won't face compiling issue. But it is quite likely that we won't get trouble with this change. :)

**Comment by fehiepsi** [comment]
@neerajprad sorry for the delay. I just update the pattern as you suggested.

File: numpyro/infer/mcmc.py [hunk 1]
```diff
@@ -437,6 +444,8 @@ class HMC(MCMCKernel):
         value is :math:`2\\pi`.
     :param callable init_strategy: a per-site initialization function.
         See :ref:`init_strategy` section for available functions.
+    :param bool find_heuristic_step_size: whether to use an extra heuristic
```
Comment (neerajprad) [inline numpyro/infer/mcmc.py:447]: Maybe we can add to the description - `"Uses a heuristic function to adjust the step size at the end of each adaptation window."`?

Also, just to confirm - this will still use this to initialize the step size but omit doing so at the end of each adaptation window?

Comment (fehiepsi) [reply numpyro/infer/mcmc.py:447]: @neerajprad We also skip the initialization of step size. Actually, this function adjusts step size at the "begin" of each adaptation window.

Comment (neerajprad) [reply numpyro/infer/mcmc.py:447]: I see, I forgot that this is a more fundamental change. I think the only time this should matter is when we are using very few warmup steps. It seems like we do this at the start and then at the end of each adaptation [window](https://github.com/pyro-ppl/numpyro/blob/master/numpyro/infer/hmc_util.py#L399)?

Comment (fehiepsi) [reply numpyro/infer/mcmc.py:447]: > the only time this should matter is when we are using very few warmup steps

Exactly!

> It seems like we do this at the start and then at the end of each adaptation window?

Actually, we do this only at the start of each adaptation window (including the first window).
